dali/operators/reader/loader/video/frames_decoder_gpu.cc

```diff
@@ -288,6 +266,110 @@ bool FramesDecoderGpu::ReadNextFrame(uint8_t *data, bool copy_to_output) {
 }
 
+bool FramesDecoderGpu::SendFrameToParser() {
+  if (av_state_->packet_->stream_index != av_state_->stream_id_) {
+    return false;
+  }
+
+  // Store pts from current packet to indicate,
+  // that this frame is in the decoder
+  piped_pts_.push(av_state_->packet_->pts);
+
+  // Add header needed for NVDECODE to the packet
+  if (filtered_packet_->data) {
+    av_packet_unref(filtered_packet_);
+  }
+  DALI_ENFORCE(av_bsf_send_packet(bsfc_, av_state_->packet_) >= 0);
+  DALI_ENFORCE(av_bsf_receive_packet(bsfc_, filtered_packet_) >= 0);
+
+  // Prepare nv packet
+  CUVIDSOURCEDATAPACKET *packet = &nvdecode_state_->packet;
+  memset(packet, 0, sizeof(CUVIDSOURCEDATAPACKET));
+  packet->payload = filtered_packet_->data;
+  packet->payload_size = filtered_packet_->size;
+  packet->flags = CUVID_PKT_TIMESTAMP;
+  packet->timestamp = filtered_packet_->pts;
+
+  // Send packet to the nv decoder
+  frame_returned_ = false;
+  CUDA_CALL(cuvidParseVideoData(nvdecode_state_->parser, packet));
+  return true;
+}
+
+bool FramesDecoderGpu::ReadNextFrameWithoutIndex(uint8_t *data, bool copy_to_output) {
+  current_copy_to_output_ = copy_to_output;
+  current_frame_output_ = data;
+
+  // Initial fill of the buffer
+  while (HasEmptySlot() && more_frames_to_decode_) {
+    if (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
+      if (!SendFrameToParser()) {
+        continue;
+      }
+    } else {
+      SendLastPacket();
+      more_frames_to_decode_ = false;
+    }
+  }
+
+  int frame_to_return_index = -1;
+  for (size_t i = 0; i < frame_buffer_.size(); ++i) {
+    if (frame_buffer_[i].pts_ != -1) {
+      frame_to_return_index = i;
+      break;
+    }
+  }
+
+  for (size_t i = 1; i < frame_buffer_.size(); ++i) {
+    if (frame_buffer_[i].pts_ != -1) {
+      if (frame_buffer_[frame_to_return_index].pts_ > frame_buffer_[i].pts_) {
+        frame_to_return_index = i;
+      }
+    }
+  }
+
+  copyD2D(
+    current_frame_output_,
+    frame_buffer_[frame_to_return_index].frame_.data(),
+    FrameSize());
+  LOG_LINE << "Read frame, index " << next_frame_idx_ << ", timestamp " <<
+          std::setw(5) << frame_buffer_[frame_to_return_index].pts_ <<
+          ", current copy " << copy_to_output << std::endl;
+  ++next_frame_idx_;
+
+  frame_buffer_[frame_to_return_index].pts_ = -1;
+
+  if (EmptyBuffer()) {
+    next_frame_idx_ = -1;
+  }
+
+  while (HasEmptySlot() && more_frames_to_decode_) {
+    if (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
+      if (SendFrameToParser()) {
+        continue;
+      }
+    } else {
+      SendLastPacket();
+      more_frames_to_decode_ = false;
+    }
+  }
+
+  return true;
+}
+
+bool FramesDecoderGpu::ReadNextFrame(uint8_t *data, bool copy_to_output) {
+  // No more frames in the file
+  if (next_frame_idx_ == -1) {
+    return false;
+  }
+
+  if (HasIndex()) {
+    return ReadNextFrameWithIndex(data, copy_to_output);
+  }
+
+  return ReadNextFrameWithoutIndex(data, copy_to_output);
```

Comment:
Nitpick - for symmetry
```suggestion
  if (HasIndex()) {
    return ReadNextFrameWithIndex(data, copy_to_output);
  } else {
    return ReadNextFrameWithoutIndex(data, copy_to_output);
  }
```

Reply:
Done